### PR TITLE
Refactoring group 5

### PR DESCRIPTION
**Pull request by jboettcher**

### CODE REVIEW

File: src/main/java/flight/reservation/plane/Helicopter.java [hunk 1]
```diff
@@ -1,25 +1,30 @@
...
 
-    public Helicopter(String model) {
-        this.model = model;
-        if (model.equals("H1")) {
-            passengerCapacity = 4;
-        } else if (model.equals("H2")) {
-            passengerCapacity = 6;
-        } else {
-            throw new IllegalArgumentException(String.format("Model type '%s' is not recognized", model));
+    public enum HeliModel {
```
Comment (delsner) [inline src/main/java/flight/reservation/plane/Helicopter.java:6]: nice solution

File: src/main/java/flight/reservation/payment/Paypal.java [hunk 2]
```diff
@@ -1,13 +1,52 @@
...
+
+
+   public Paypal(String email, String password)  {
+       super();
+       this.email = email;
+       this.password = password;
+   }
+
+
+    public boolean processOrder(FlightOrder flightOrder) throws IllegalStateException {
```
Comment (delsner) [inline src/main/java/flight/reservation/payment/Paypal.java:27]: the `processOrder()` method could contain less duplicate code by factoring out the duplicate code

File: src/main/java/flight/reservation/payment/PaymentMethod.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
+package flight.reservation.payment;
+
+import flight.reservation.order.FlightOrder;
+
+import java.util.Date;
+
+public abstract class PaymentMethod {
+    protected double amount;
+
+    public PaymentMethod() {
```
Comment (delsner) [inline src/main/java/flight/reservation/payment/PaymentMethod.java:10]: constructor could be removed and `amount` directly initialized with 0.

File: src/main/java/flight/reservation/plane/PassengerPlane.java [hunk 4]
```diff
@@ -1,33 +1,46 @@
 package flight.reservation.plane;
 
-public class PassengerPlane {
+public class PassengerPlane extends Aircraft {
 
-    public String model;
-    public int passengerCapacity;
-    public int crewCapacity;
+    public enum PlaneModel {
```
Comment (delsner) [inline src/main/java/flight/reservation/plane/PassengerPlane.java:5]: again, nicely solved

File: src/main/java/flight/reservation/payment/CreditCard.java [hunk 5]
```diff
@@ -1,31 +1,55 @@
...
-        return amount;
+    public boolean pay(FlightOrder flightOrder, double amount) throws IllegalStateException {
+        if (isValid()) {
+            System.out.println("Paying " + flightOrder.getPrice() + " using Credit Card.");
+            double remainingAmount = getAmount() - flightOrder.getPrice();
+            if (remainingAmount < 0) {
+                System.out.printf("Card limit reached - Balance: %f%n", remainingAmount);
+                throw new IllegalStateException("Card limit reached");
+            }
+            setAmount(remainingAmount);
```
Comment (delsner) [inline src/main/java/flight/reservation/payment/CreditCard.java:48]: not sure if `public setAmount()` is necessary? Here, you could directly operate on the `protected` member variable.

File: src/test/java/flight.reservation/ScenarioTest.java [hunk 6]
```diff
@@ -199,8 +203,8 @@ public void initCreditCard() {
             @DisplayName("then the booking should succeed")
             void thenTheBookingShouldSucceed() throws NoSuchFieldException {
                 ScheduledFlight scheduledFlight = schedule.searchScheduledFlight(flight.getNumber());
-                FlightOrder order = customer.createOrder(Arrays.asList("Max"), Arrays.asList(scheduledFlight), 100);
-                boolean processed = order.processOrderWithCreditCard(creditCard);
+                FlightOrder order = customer.createOrder(Collections.singletonList("Max"), Collections.singletonList(scheduledFlight), 100);
+                boolean processed = creditCard.processOrder(order);
```
Comment (delsner) [inline src/test/java/flight.reservation/ScenarioTest.java:207]: apparently you have changed the existing API of `FlightOrder` (might be a problem for other dependent code)

File: src/main/java/flight/reservation/payment/Paypal.java [hunk 7]
```diff
@@ -1,13 +1,52 @@
...
+            // Payment is already proceeded
+            return true;
+        }
+        // validate payment information
+        if (email == null || password == null || !email.equals(DATA_BASE.get(password))) {
+            throw new IllegalStateException("Payment information is not set or not valid.");
+        }
+        boolean isPaid = pay(flightOrder, flightOrder.getPrice());
+        if (isPaid) {
+            flightOrder.setClosed();
```
Comment (heidekrueger) [inline src/main/java/flight/reservation/payment/Paypal.java:38]: Here you introduce a side effect on the `order` object.

**Pull request closed without merging**